lib/saluki-components/src/sources/checks/mod.rs

```diff
@@ -0,0 +1,254 @@
+use saluki_core::{
+    components::{sources::*, ComponentContext},
+    topology::{
+        shutdown::{DynamicShutdownCoordinator, DynamicShutdownHandle},
+        OutputDefinition,
+    },
+};
+use saluki_env::autodiscovery::{AutodiscoveryEvent, AutodiscoveryProvider, Config};
+use saluki_error::{generic_error, GenericError};
+use saluki_event::DataType;
+use saluki_health::Health;
+use serde::Deserialize;
+use tokio::select;
+use tokio::sync::broadcast::Receiver;
+use tracing::{debug, error, info, warn};
+
+mod scheduler;
+use self::scheduler::Scheduler;
+
+mod check;
+use self::check::Check;
+
+const fn default_check_runners() -> usize {
+    4
+}
+
+/// Configuration for the checks source.
+#[derive(Deserialize)]
+pub struct ChecksConfiguration {
+    /// The number of check runners to use.
+    ///
+    /// Defaults to 4.
+    #[serde(default = "default_check_runners")]
+    check_runners: usize,
+
+    /// Autodiscovery provider to use.
+    #[serde(skip)]
+    autodiscovery_provider: Option<Arc<dyn AutodiscoveryProvider + Send + Sync>>,
+}
+
+impl ChecksConfiguration {
+    /// Creates a new `ChecksConfiguration` from the given configuration.
+    pub fn from_configuration(config: &GenericConfiguration) -> Result<Self, GenericError> {
+        Ok(config.as_typed()?)
+    }
+
+    /// Sets the autodiscovery provider to use.
+    pub fn with_autodiscovery_provider<A>(mut self, provider: A) -> Self
+    where
+        A: AutodiscoveryProvider + Send + Sync + 'static,
+    {
+        self.autodiscovery_provider = Some(Arc::new(provider));
+        self
+    }
+}
+
+#[async_trait]
+impl SourceBuilder for ChecksConfiguration {
+    async fn build(&self, _context: ComponentContext) -> Result<Box<dyn Source + Send>, GenericError> {
+        match &self.autodiscovery_provider {
+            Some(autodiscovery) => Ok(Box::new(ChecksSource {
+                autodiscovery: Arc::clone(autodiscovery),
+                check_runners: self.check_runners,
+            })),
+            None => Err(generic_error!("No autodiscovery provider configured.")),
+        }
+    }
+
+    fn outputs(&self) -> &[OutputDefinition] {
+        static OUTPUTS: [OutputDefinition; 1] = [OutputDefinition::default_output(DataType::Metric)];
+
+        &OUTPUTS
+    }
+}
+
+impl MemoryBounds for ChecksConfiguration {
+    fn specify_bounds(&self, builder: &mut MemoryBoundsBuilder) {
+        builder.minimum().with_single_value::<ChecksSource>("component struct");
+    }
+}
+
+struct ChecksSource {
+    autodiscovery: Arc<dyn AutodiscoveryProvider + Send + Sync>,
+    check_runners: usize,
+}
+
+trait CheckBuilder {
+    fn build_check(&self, check_id: &str, check_request: &Config) -> Option<Arc<dyn Check + Send + Sync>>;
+}
+
+/// A no-op implementation of CheckBuilder
+struct NoOpCheckBuilder;
+
+impl CheckBuilder for NoOpCheckBuilder {
+    fn build_check(&self, _check_id: &str, _check_request: &Config) -> Option<Arc<dyn Check + Send + Sync>> {
+        None
+    }
+}
```

Comment:
We should create a new module -- `lib/saluki-components/src/sources/checks/builder` -- and put this stuff under there.